Fix corner starts

TomMelt opened the pull request:
This PR fixes corner neighbour metadata for periodic boundaries

## Key Changes

- Fixed `haloCornerBufferPositions()` logic to correctly handle periodic boundary conditions
- Updated corner neighbour detection to properly identify neighbours across periodic boundaries
- Added comprehensive tests for halo buffer positions and corner neighbour detection
- Updated reference test data for integration tests
- Improved documentation with better comments and docstrings (but more to come in #80)

## Tests

* `test_haloBufferPositions.cpp`
* `test_haloCornerBufferPositions.cpp`
* `test_is_corner_neighbour.cpp`
* `test_is_neighbour.cpp`

TODO:

- [x] add unit tests for `haloCornerBufferPositions`
- [x] add unit tests for `haloBufferPositions` functions @niravshah241 is working on
- [x] rename `haloBufferPositions` to `haloEdgeBufferPositions` (this will be handled in a following PR -- see issue #80 )
- [x] fix `haloCornerBufferPositions` for periodic corner neighbours (@niravshah241 is double checking) 
- [x] @niravshah241 is investigating why the `test_halo_corner_start` test is failing -> See https://github.com/
nextsimhub/domain_decomp/pull/78/commits/b8ef5f59ec36814b6029fe174bc70a37ba888581
- [x] @niravshah241 to resolve merge conflict
- [x] Add test diagrams in repository for easier reference later @niravshah241 
- [x] check all edge metadata in `haloBufferPositions` (i.e., use same approach as `haloCornerBufferPositions`)

TomMelt committed:
wip: initial fix for non-periodic neighbours

niravshah241 commented on Partitioner.cpp:
```diff
@@ -155,64 +155,57 @@ void Partitioner::haloBufferPositions(
-            start = left - d2.p1.x - 1;
+            // this second case works if the corner neighbour lies on the bottom edge (or in the
+            // corner of both e.g., the bottom left corner)
+            int dx = d1.p2.y - d2.p1.y;
```
Should this be `dx = d1.p2.x - d2.p1.x`?

TomMelt replied:
good point. I have either named `dx` incorrectly, or it should be `d1.p2.x`. I will double check these and get back to you

niravshah241 commented on Partitioner.cpp:
```diff
@@ -155,64 +155,57 @@ void Partitioner::haloBufferPositions(
+            send_pos = d2.get_width() + dy;
         } else {
-            start = (top - d2.p1.y) * d2.get_width();
+            int dx = d1.p2.y - d2.p1.y;
```
Should this be `dx = d2.p1.x - d1.p1.x - 1`?

TomMelt replied:
same as prev comment.

niravshah241 commented on Partitioner.cpp:
```diff
@@ -155,64 +155,57 @@ void Partitioner::haloBufferPositions(
+            send_pos = d2.get_width() + dy - 1;
         } else {
-            start = (bottom - d2.p1.y - 1) * d2.get_width();
+            int dx = d1.p2.y - d2.p1.y;
```
Should this be `dx=d1.p1.x-d2.p1.x`?

TomMelt replied:
same as prev comment.

niravshah241 commented on Partitioner.cpp:
```diff
@@ -155,64 +155,57 @@ void Partitioner::haloBufferPositions(
+            send_pos = 2 * d2.get_width() + d2.get_height() + dy - 1;
         } else {
-            start = (bottom - d2.p1.y) * d2.get_width() - 1;
+            int dx = d1.p2.y - d2.p1.y;
```
Should this be `dx = d1.p2.x - d2.p1.x` ?

TomMelt replied:
same as prev comment.

niravshah241 commented on Partitioner.hpp:
Do we necessarily need `private` here? If no, it would be better (at least for the unit tests) if the functions `is_neighbour` , `is_corner_neighbour`, `haloBufferPositions` and `haloCornerBufferPositions` are made public.

niravshah241 replied:
Made ```public``` from ```private```. See https://github.com/nextsimhub/domain_decomp/commit/b8ef5f59ec36814b6029fe174bc70a37ba888581 .

TomMelt replied:
Thanks @niravshah241

niravshah241 reviewed (commented):
Overall, looks good. Some quick formulae checks and making `private` members to `public` in `partition.hpp` have been suggested.

The unit tests for `is_corner_neighbour` and `haloCornerBufferPositions` (previously `halo_corner_start`) have been written in the branch `add-periodic-corner-nbrs`. (It will need some changes depending upon final function calls and index changes.)

TomMelt committed:
feat: fix corner neighbour for periodic BC

TomMelt committed:
test: update test reference data

TomMelt committed:
wip: Nirav see if you can find why its breaking

Commit:
Tests updated and fixes in Partition.cpp

niravshah241 commented:
Tests are added to this branch. The tests in `test_halo_corner_start` nonperiodic cases are fixed now https://github.com/nextsimhub/domain_decomp/pull/78/commits/b8ef5f59ec36814b6029fe174bc70a37ba888581 .

Commit:
DOmain decomp test in works

niravshah241 commented:
Tests related to `is_neighbour` and `haloBufferPositions` added. There is difference in how the neighbours are considered in two functions. Given 2 subdomains `d1` and `d2`, the `RIGHT` in `is_neighbour` means `d2` is `RIGHT` of `d1` but in `haloBufferPositions`, `d1` is considered `RIGHT` of `d2`. Also, for `LEFT`, `BOTTOM` and `TOP` .

In addition the images for tests have now been added in `img` folder.

Commit:
Periodic tests added

TomMelt committed:
doc: correct typo in comment regarding haloBufferPositions

TomMelt committed:
feat: fix haloCornerBufferPositions logic

TomMelt committed:
test: generalize test_haloCornerBufferPositions.cpp

TomMelt committed:
test: update reference data for integration test

TomMelt commented:
> Tests related to `is_neighbour` and `haloBufferPositions` added. There is difference in how the neighbours are considered in two functions. Given 2 subdomains `d1` and `d2`, the `RIGHT` in `is_neighbour` means `d2` is `RIGHT` of `d1` but in `haloBufferPositions`, `d1` is considered `RIGHT` of `d2`. Also, for `LEFT`, `BOTTOM` and `TOP` .
> 
> In addition the images for tests have now been added in `img` folder.

There is no difference. I had made a typo in the doc string but the functionality is correct. This has been fixed in commit https://github.com/nextsimhub/domain_decomp/pull/78/commits/5380006efe2805f9a38084d429203060ca09ffb0

Please remove the image from the img folder. If we want to upload an image it should be svg and have more description. FWIW we can see the domain layout looking at the partition_mask files.

Commit:
test_haloBufferPositions updated

Commit:
domain decomp image

joewallwork commented on test/CMakeLists.txt:
```diff
@@ -32,8 +32,10 @@ create_test(TARGET test_grid_2 SOURCES test_grid_2.cpp)
+create_test(TARGET test_haloCornerBufferPositions SOURCES test_haloCornerBufferPositions.cpp)
 create_test(TARGET test_is_corner_neighbour SOURCES test_is_corner_neighbour.cpp)
+create_test(TARGET test_is_neighbour SOURCES test_is_neighbour.cpp)
+create_test(TARGET test_haloBufferPositions SOURCES test_haloBufferPositions.cpp)
```
Very minor point: you could consider using a consistent case convention for naming the files. Probably best to make it consistent with nextSIM-DG.

TomMelt replied:
Thanks.... This is on my todo list already :sweat: I made issue #80

I will make it a separate PR to address though.

joewallwork commented on Partitioner.hpp:
```diff
@@ -249,50 +243,34 @@ class LIB_EXPORT Partitioner {
-     *  │ └───────────┴─────┘ └──────────┴────┘  └──────────┴────┘
-     *  │0            5     7
-     *  └───►X
+     * @brief Compute the sendPos and recvPos for halo exchange.
```
Suggest spelling this out because it might not be immediately obvious what `sendPos` and `recvPos` mean.

TomMelt replied:
Thanks @joewallwork I think it got lost in the diff. But I had added description to the vars below:

https://github.com/nextsimhub/domain_decomp/blob/8b98a5870b2c7d895897418c546afbdadad0baeb/Partitioner.hpp#L252-L261

Do you want more explanation or is it ok as is?

joewallwork replied:
Ah okay got it, thanks.

joewallwork commented on Partitioner.cpp:
```diff
@@ -17,6 +17,27 @@
+/*!
+ * @brief Compute modulo operation with non-negative result
+ *
+ * This function computes a % n and ensures the result is always non-negative.
```
Worth spelling out `a % n` because it might not be obvious. This isn't the same notation that would be used in maths, for example.

TomMelt replied:
Thanks @joewallwork done in e23165f

joewallwork commented on test/test_haloEdgeBufferPositions.cpp:
Do I recall correctly that we decided to use the term `haloEdgeBuffer` to distinguish from corner case?

TomMelt replied:
yes this will also be done as part of a separate PR fixing #80

joewallwork commented on test/test_haloCornerBufferPositions.cpp:
```diff
@@ -0,0 +1,126 @@
+/*!
+ * @file halo_corner_start.cpp
```
Name is inconsistent with filename.

TomMelt replied:
thanks done in 8b98a58

joewallwork commented on test/test_haloBufferPositions.cpp:
```diff
@@ -0,0 +1,143 @@
+extern int global_argc;
+extern char** global_argv;
+
+MPI_TEST_CASE("Corner neighbour: Non-periodic, 4 MPI ranks", 4)
```
This isn't really true - is it not edge neighbours? Also, the periodic case is also considered in this test as well as the non-periodic case.

TomMelt replied:
thanks done in a81b85c

joewallwork commented on test/test_haloCornerBufferPositions.cpp:
```diff
@@ -0,0 +1,126 @@
+extern int global_argc;
+extern char** global_argv;
+
+MPI_TEST_CASE("Corner neighbour: Non-periodic, 4 MPI ranks", 4)
```
Again, the periodic case is also considered in this test as well as the non-periodic case.

TomMelt replied:
thanks done in a81b85c

joewallwork commented on test/test_is_neighbour.cpp:
```diff
@@ -0,0 +1,123 @@
+/*!
+ * @file .cpp
```
Missing filename

TomMelt replied:
thanks done in 8b98a58

joewallwork commented on test/test_is_neighbour.cpp:
```diff
@@ -0,0 +1,123 @@
+    if (test_rank == 0) {
+        neighbour = partitioner->is_neighbour(domains[0], domains[2], RIGHT);
+        REQUIRE(neighbour);
+    }
```
Could we not just do something like this? (And similarly elsewhere.)
```suggestion
    // RIGHT
    if (test_rank == 0) {
        REQUIRE(partitioner->is_neighbour(domains[0], domains[2], RIGHT));
    }
```

TomMelt replied:
I just realised that the testing of `partitioner->is_neighbour` is now much better handled in `test_haloBufferPositions.cpp`. So I would prefer to remove this file altogether.

Sorry for the additional review load :grimacing: 

My problem with this specific file is it doesn't check all cases. Just one for each rank. Whereas the `test_haloBufferPositions.cpp` is designed to check all of them.

joewallwork replied:
Sure, no worries. Will re-review now.

joewallwork reviewed (changes requested):
Thanks for this. I've not gone through all the working of the subdomains, numbers, neighbours etc because I figured you'd have it right between the two of you. However, I do have some comments and suggestions that it'd be good to address before merging.

TomMelt commented on test/test_haloBufferPositions.cpp:
```diff
@@ -0,0 +1,143 @@
+    // The int in tuple<Edge, int> refers to rank. In the case of edge neighbours,
+    // it is possible that one edge may corresspond to multiple neighbour unlike
+    // corner neighbour where a given vertice can correspond to only one neighbour.
+    std::map<std::tuple<Edge, int>, HaloEdgeInfo> edgeInfoExpected, edgeInfoActual;
```
```suggestion
    std::map<Edge>, std::list<HaloEdgeInfo>> edgeInfoExpected, edgeInfoActual;
```

The order of this data structure is slightly wrong. We really want a map of Edges with the respective neighbor data as shown above.

It should look something like this:

```
{
 'RIGHT': (haloEdgeInfo1, haloEdgeInfo2, haloEdgeInfo3),
  'LEFT': (haloEdgeInfo1),
   'TOP': (haloEdgeIhaloEdgeInfofo1, haloEdgeInfo2, haloEdgeInfo3, haloEdgeInfo4),
'BOTTOM': (haloEdgeInfo1, haloEdgeInfo2)
}
```

I would also recommend changing from a tuple to a list because you will need to append to it for `edgeInfoActual`

TomMelt commented on test/test_haloEdgeBufferPositions.cpp:
```diff
@@ -0,0 +1,143 @@
+
+    int start;
+    int recv;
+
+    struct HaloEdgeInfo {
+        bool isPeriodic;
+        int startPos;
+        int recvPos;
+    };
```
```suggestion
    struct HaloEdgeInfo {
        bool isPeriodic;
        int rank;
        int startPos;
        int recvPos;
    };
```

You need to extend this object `HaloEdgeInfo` to add the rank (as done in the corner test).

TomMelt commented on test/test_haloBufferPositions.cpp:
```diff
@@ -0,0 +1,143 @@
+            // periodic neighbours
+            if (partitioner->is_neighbour(domains[test_rank], domains[p], edge, pxOn, pyOn)) {
+                partitioner->haloBufferPositions(domains[test_rank], domains[p], edge, start, recv);
+                edgeInfoActual[{ edge, p }] = { true, start, recv };
```
You will need to update this logic, to append the neighbor info one at a time.

Once you have changed `edgeInfoExpected` and `edgeInfoActual` to lists, you will be able to append.

It may look something like, you would need to check it exists before you try to append for example. 

```suggestion
                edgeInfoActual.at(edge).append(true, p, start, recv);
```

First check it exists, if yes then append. If no, then create a list with 1 item and store it in the map for that specific edge.

Commit:
is_neighbour test fixed, corrections in is_neighbour function

TomMelt committed:
chore: use std::vector instead of list

TomMelt committed:
chore: move edge neighbour check logic to is_neighbour

remove neighbour check logic from domain_overlap

update comments in is_neighbour to reflect the changes

in Partitioner we still check the halo_size is non-zero but this will
always be true, so it is not a problem.

TomMelt committed:
chore: update doc for mod function

TomMelt committed:
chore: update filenames in header

TomMelt committed:
test: remove redundant tests

TomMelt committed:
chore: remove unused vars

TomMelt committed:
chore: tidy buffer tests

TomMelt committed:
chore: rename haloBufferPositions to haloEdgeBufferPositions

TomMelt committed:
chore: rename vertices to corners

TomMelt committed:
chore: rename vars to camelCase to match nextsim

joewallwork commented on Partitioner.cpp:
```diff
@@ -155,64 +179,64 @@ void Partitioner::haloBufferPositions(
+        // for a TOP_RIGHT vertex the corner neighbour can either be along the other domains bottom
+        // or left edge. We need to check so we know where to look in the send buffer.
+        xpos = mod((d1.p2.x), globalX);
+        ypos = mod((d1.p2.y), globalY);
```
```suggestion
        xpos = mod(d1.p2.x, globalX);
        ypos = mod(d1.p2.y, globalY);
```

joewallwork commented on Partitioner.cpp:
```diff
@@ -155,64 +179,64 @@ void Partitioner::haloBufferPositions(
-            start = right - d2.p1.x;
+    } else if (vertex == TOP_LEFT) {
+        xpos = mod((d1.p1.x - 1), globalX);
+        ypos = mod((d1.p2.y), globalY);
```
```suggestion
        xpos = mod(d1.p1.x - 1, globalX);
        ypos = mod(d1.p2.y, globalY);
```

joewallwork commented on Partitioner.cpp:
```diff
@@ -155,64 +179,64 @@ void Partitioner::haloBufferPositions(
-            start = (bottom - d2.p1.y - 1) * d2.get_width() + (right - d2.p1.x);
+    } else if (vertex == BOTTOM_LEFT) {
+        xpos = mod((d1.p1.x - 1), globalX);
+        ypos = mod((d1.p1.y - 1), globalY);
```
```suggestion
        xpos = mod(d1.p1.x - 1, globalX);
        ypos = mod(d1.p1.y - 1, globalY);
```

joewallwork commented on Partitioner.cpp:
```diff
@@ -155,64 +179,64 @@ void Partitioner::haloBufferPositions(
-            start = (bottom - d2.p1.y - 1) * d2.get_width() + (left - d2.p1.x) - 1;
+    } else if (vertex == BOTTOM_RIGHT) {
+        xpos = mod((d1.p2.x), globalX);
+        ypos = mod((d1.p1.y - 1), globalY);
```
```suggestion
        xpos = mod(d1.p2.x, globalX);
        ypos = mod(d1.p1.y - 1, globalY);
```

TomMelt replied:
Thanks @joewallwork for the suggestions. I have applied them all 👌

joewallwork reviewed (approved):
I just have some minor formatting suggestions but otherwise I think it looks good.

TomMelt committed:
chore: apply formatting changes

Co-authored-by: Joe Wallwork <22053413+joewallwork@users.noreply.github.com>
